File: backend/src/main/scala/domains/user/api/CreateUserProfileSettings.scala
package domains.user.api

import cats.effect.IO
import domains.auth.api.InternalOnlyApi
import domains.problem.objects.ProblemTitleDisplayMode
import domains.user.objects.{DisplayName, UserDisplayMode, UserLocale, Username}
import domains.user.objects.UserProfileSettings
import domains.user.table.user_profile.UserProfileTable
import org.http4s.Method
import shared.api.ApiPath

import java.sql.Connection

/** 内部用户资料创建 API，通常由注册流程在创建账号后调用。 */
object CreateUserProfileSettings extends InternalOnlyApi[CreateUserProfileSettings.Input, UserProfileSettings]:

  /** 创建用户资料所需的完整输入，字段直接映射 user_profiles 表。 */
  final case class Input(
    username: Username,
    displayName: DisplayName,
    displayMode: UserDisplayMode,
    locale: UserLocale,
    problemTitleDisplayMode: ProblemTitleDisplayMode,
    autoMarkMessageRead: Boolean
  )

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/internal/users/profile-settings")

  /** 在当前事务中插入用户资料并返回初始设置，外部 HTTP 不可直接调用。 */
  override def plan(connection: Connection, input: Input): IO[UserProfileSettings] =
    UserProfileTable.insertProfile(
      connection,
      username = input.username,
      displayName = input.displayName,
      displayMode = input.displayMode,
      locale = input.locale,
      problemTitleDisplayMode = input.problemTitleDisplayMode,
      autoMarkMessageRead = input.autoMarkMessageRead
    )

File: frontend/src/apis/user/CreateUserProfileSettings.ts
import type { APIMessage } from '@/system/api/api-message'
import type { ProblemTitleDisplayMode } from '@/objects/problem/ProblemTitleDisplayMode'
import type { DisplayName } from '@/objects/user/DisplayName'
import type { UserDisplayMode } from '@/objects/user/UserDisplayMode'
import type { UserLocale } from '@/objects/user/UserLocale'
import type { Username } from '@/objects/user/Username'
import type { UserProfileSettings } from '@/objects/user/UserProfileSettings'

/** 内部创建用户资料设置请求体；用于账号创建后初始化资料和偏好。 */
type CreateUserProfileSettingsBody = {
  username: Username
  displayName: DisplayName
  displayMode: UserDisplayMode
  locale: UserLocale
  problemTitleDisplayMode: ProblemTitleDisplayMode
  autoMarkMessageRead: boolean
}

/** 创建用户资料设置的内部 API；输出可用于设置页的完整资料设置快照。 */
export class CreateUserProfileSettings implements APIMessage<UserProfileSettings> {
  declare readonly responseType?: UserProfileSettings
  readonly method = 'POST'
  readonly apiPath = 'internal/users/profile-settings'
  private readonly request: CreateUserProfileSettingsBody

  constructor(request: CreateUserProfileSettingsBody) {
    this.request = request
  }

  body(): CreateUserProfileSettingsBody {
    return this.request
  }
}
